<a href="https://colab.research.google.com/github/Jihyesh/codestates_projects_section1/blob/main/AI_15_%ED%99%8D%EC%A7%80%ED%98%9C_Section1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
from google.colab import files
uploaded = files.upload()

Saving vgames2.csv to vgames2.csv


In [3]:
df = pd.read_csv('vgames2.csv')
df.head()

Unnamed: 0                                            Name Platform  \
0           1                    Candace Kane's Candy Factory       DS   
1           2                                  The Munchables      Wii   
2           3  Otome wa Oanesama Boku ni Koi Shiteru Portable      PSP   
3           4                Deal or No Deal: Special Edition       DS   
4           5       Ben 10 Ultimate Alien: Cosmic Destruction      PS3   

     Year      Genre           Publisher NA_Sales EU_Sales JP_Sales  \
0  2008.0     Action           Destineer     0.04        0        0   
1  2009.0     Action  Namco Bandai Games     0.17        0        0   
2  2010.0  Adventure           Alchemist        0        0     0.02   
3  2010.0       Misc           Zoo Games     0.04        0        0   
4  2010.0   Platform         D3Publisher     0.12     0.09        0   

  Other_Sales  
0           0  
1        0.01  
2           0  
3           0  
4        0.04

##Mission
<font color='blue'>비데이터 직군</font>의 청자에게 데이터를 기반으로 위의 주제에 대해 설명하는 영상 제출
- **지역에 따라서 선호하는 게임 장르가 다를까** 라는 질문에 대답을 하셔야 합니다
- **연도별 게임의 트렌드가 있을까** 라는 질문에 대답을 하셔야 합니다
- **출고량의 높은 게임에 대한 분석 및 시각화 프로세스**가 포함되어야 합니다
- **결론적으로 다음 분기에 어떤 게임을 설계해야 하는지를 제시**하셔야 합니다
- 추가 인사이트


##EDA

In [ ]:
df.shape

(16598, 10)

In [ ]:
df.dtypes

Unnamed: 0       int64
Name            object
Platform        object
Year           float64
Genre           object
Publisher       object
NA_Sales        object
EU_Sales        object
JP_Sales        object
Other_Sales     object
dtype: object

In [ ]:
df.sample(10)

Unnamed: 0                                             Name Platform  \
5988         5989         Jimmie Johnson's Anything With an Engine      Wii   
15253       15254                                    NBA Live 2005      PS2   
13037       13038                                    Vampire Night      PS2   
11765       11766                                            Skate     X360   
4901         4902                         Chaotic: Shadow Warriors     X360   
14044       14045                              Colin McRae Rally 3      PS2   
8751         8752  Final Fantasy Crystal Chronicles: Ring of Fates       DS   
9250         9251                                       Madagascar      GBA   
15459       15460                            Gekka Ryouran Romance      PSP   
7278         7279                           Divinity: Original Sin     XOne   

         Year         Genre                     Publisher NA_Sales EU_Sales  \
5988   2011.0        Racing  Konami Digital Entertainment     0.07        0   
15253  2004.0        Sports               Electronic Arts     2.03     0.21   
13037  2001.0       Shooter   Sony Computer Entertainment     0.07     0.05   
11765  2007.0        Sports               Electronic Arts     0.62     0.04   
4901   2009.0        Action                    Activision     0.14        0   
14044  2002.0        Racing                   Codemasters     0.03     1.05   
8751   2007.0  Role-Playing                   Square Enix     0.22     0.01   
9250   2005.0      Platform                    Activision     0.62     0.24   
15459  2011.0     Adventure                  Idea Factory        0        0   
7278   2015.0  Role-Playing        Focus Home Interactive     0.04     0.02   

      JP_Sales Other_Sales  
5988         0        0.01  
15253        0        0.35  
13037        0        0.02  
11765        0        0.06  
4901         0        0.01  
14044     0.02         0.3  
8751      0.42        0.02  
9250         0        0.03  
15459     0.02           0  
7278         0        0.01

In [ ]:
df.isna().sum()

Unnamed: 0       0
Name             0
Platform         0
Year           271
Genre           50
Publisher       58
NA_Sales         0
EU_Sales         0
JP_Sales         0
Other_Sales      0
dtype: int64

In [ ]:
df[df.duplicated()]

Empty DataFrame
Columns: [Unnamed: 0, Name, Platform, Year, Genre, Publisher, NA_Sales, EU_Sales, JP_Sales, Other_Sales]
Index: []

In [ ]:
df.describe()

Unnamed: 0          Year
count  16598.000000  16327.000000
mean    8299.500000   1994.242543
std     4791.574219    155.077168
min        1.000000      0.000000
25%     4150.250000   2003.000000
50%     8299.500000   2007.000000
75%    12448.750000   2010.000000
max    16598.000000   2020.000000

###EDA-Year

In [ ]:
df.Year.dtype

dtype('float64')

In [ ]:
# Year missing value
df[df.Year.isnull()]

Unnamed: 0                                         Name Platform  Year  \
31             32                                    Rock Band     X360   NaN   
109           110   Yu-Gi-Oh! 5D's Wheelie Breakers (JP sales)      Wii   NaN   
273           274  Jewel Link Chronicles: Mountains of Madness       DS   NaN   
358           359           Majesty 2: The Fantasy Kingdom Sim     X360   NaN   
430           431                                   Famista 64      N64   NaN   
...           ...                                          ...      ...   ...   
16240       16241                                Freaky Flyers       GC   NaN   
16275       16276                               Space Invaders     2600   NaN   
16278       16279                                Madden NFL 11      Wii   NaN   
16425       16426                               Yoostar on MTV     X360   NaN   
16432       16433                                  Bejeweled 3     X360   NaN   

            Genre           Publisher NA_Sales EU_Sales JP_Sales Other_Sales  
31           Misc     Electronic Arts     1.93     0.34        0        0.21  
109        Racing             Unknown        0        0     0.02           0  
273        Puzzle           Avanquest        0     0.06        0        0.01  
358    Simulation             Unknown     0.03        0        0           0  
430        Sports  Namco Bandai Games        0        0     0.17        0.03  
...           ...                 ...      ...      ...      ...         ...  
16240      Racing             Unknown     0.01        0        0           0  
16275     Shooter               Atari     2.36     0.14       0M        0.03  
16278      Sports             Unknown      0.7        0        0         50K  
16425        Misc             Unknown     0.06        0        0           0  
16432      Puzzle             Unknown     0.13        0        0        0.01  

[271 rows x 10 columns]

In [ ]:
df[df.Year.isnull()].shape[0]/df.shape[0]

0.016327268345583804

###EDA-Platform

In [ ]:
df[df.Platform.isnull()]

Empty DataFrame
Columns: [Unnamed: 0, Name, Platform, Year, Genre, Publisher, NA_Sales, EU_Sales, JP_Sales, Other_Sales]
Index: []

In [ ]:
df.Platform.unique()

array(['DS', 'Wii', 'PSP', 'PS3', 'PC', 'PS', 'GBA', 'PS4', 'PS2', 'XB',
       'X360', 'GC', '3DS', '2600', 'SAT', 'GB', 'NES', 'DC', 'N64',
       'XOne', 'SNES', 'WiiU', 'PSV', 'GEN', 'SCD', 'WS', 'NG', 'TG16',
       '3DO', 'GG', 'PCFX'], dtype=object)

In [ ]:
df.Platform.nunique()

31

###EDA-Genre

In [ ]:
df[df.Genre.isnull()].shape[0]

50

In [ ]:
df[df.Genre.isnull()]

Unnamed: 0                                         Name Platform  \
50             51                     Ape Escape: On the Loose      PSP   
408           409                                  NBA Ballers       XB   
1328         1329                                   NFL Street       XB   
1582         1583               SingStar Singalong With Disney      PS2   
1971         1972                          Pokemon X/Pokemon Y      3DS   
2043         2044                                NASCAR Racing       PS   
2497         2498             Jurassic Park III: Island Attack      GBA   
3353         3354                  The Godfather: Dons Edition      PS3   
3363         3364                 Ski-Doo Snowmobile Challenge     X360   
3470         3471           Batman Beyond: Return of the Joker      N64   
3515         3516     Princess Saver! Eternal Love for My Lady      PS2   
3932         3933                            MVP Baseball 2004      PS2   
4148         4149                          Grand Theft Auto IV       PC   
4349         4350                    Brain Quest: Grades 5 & 6       DS   
4545         4546            Dragon Ball Z: Ultimate Tenkaichi     X360   
4598         4599                                    Wii Party      Wii   
4935         4936                             Astonishia Story      PSP   
5668         5669                                 Chrono Cross       PS   
5713         5714                 Back to the Future: The Game     XOne   
5945         5946                            Medabots: Metabee      GBA   
6798         6799                                 UFO: Trilogy       PC   
7143         7144                 Largo Winch .// Commando Sar       PS   
7336         7337                                Madden NFL 11      PS2   
7526         7527                              Ukiyo no Shishi      PS3   
7797         7798                    Myth Makers: Orbs of Doom      Wii   
9488         9489                            Final Fantasy XII      PS2   
10432       10433        The Mummy: Tomb of the Dragon Emperor      Wii   
10501       10502                      Bakemonogatari Portable      PSP   
10874       10875            Karaoke Revolution Glee: Volume 3      Wii   
11050       11051                                 Puppy Palace       DS   
11063       11064                                     Yakuza 3      PS3   
11268       11269            Cloudy With a Chance of Meatballs      PS3   
11496       11497                                     F-Zero X      N64   
11668       11669  The History Channel: Battle for the Pacific      Wii   
11699       11700                               Time Travelers      PSV   
12314       12315                                   Ice Hockey     2600   
12331       12332                           Onimusha: Warlords      PS2   
12340       12341                             Virtual Chess 64      N64   
12870       12871         PQ2: Practical Intelligence Quotient      PSP   
12982       12983                                  Urban Chaos       PS   
13348       13349        Treasure Report: Kikai Jikake no Isan       DS   
14930       14931                            Skylanders Giants      PS3   
15077       15078                   Myth Makers: Super Kart GP      Wii   
15211       15212                    James Bond 007: Nightfire       XB   
15426       15427            WarioWare, Inc.: Mega Party Game$       GC   
15526       15527                                   Bloodborne      PS4   
16091       16092                       Brain Boost: Beta Wave       DS   
16431       16432            Cars: Mater-National Championship     X360   
16555       16556                          Apache: Air Assault       PC   
16587       16588                 Need for Speed Underground 2       XB   

         Year Genre                     Publisher NA_Sales EU_Sales JP_Sales  \
50     2005.0   NaN   Sony Computer Entertainment     0.52     0.01     0.12   
408    2004.0   NaN  

In [ ]:
df.Genre.unique()

array(['Action', 'Adventure', 'Misc', 'Platform', 'Sports', 'Simulation',
       'Racing', 'Role-Playing', 'Puzzle', 'Strategy', 'Fighting',
       'Shooter', nan], dtype=object)

In [ ]:
df[df.Genre == 'Platform']

Unnamed: 0                                       Name Platform    Year  \
4               5  Ben 10 Ultimate Alien: Cosmic Destruction      PS3  2010.0   
20             21                     Ty the Tasmanian Tiger      PS2  2002.0   
25             26                                   Dr. Muto       GC  2002.0   
28             29                    Famicom Mini: Makaimura      GBA  2004.0   
34             35            Mega Man Anniversary Collection      PS2  2004.0   
...           ...                                        ...      ...     ...   
16531       16532                  New Super Mario Bros. Wii      Wii  2009.0   
16563       16564                                     Wall-E      Wii  2008.0   
16578       16579                       Gex: Enter the Gecko       PS  1998.0   
16579       16580                     Super Ghouls 'n Ghosts      GBA  2002.0   
16593       16594                    Ice Age 2: The Meltdown       GC  2006.0   

          Genre                      Publisher NA_Sales EU_Sales JP_Sales  \
4      Platform                    D3Publisher     0.12     0.09        0   
20     Platform                Electronic Arts     0.57     0.45        0   
25     Platform                   Midway Games     0.02     0.01        0   
28     Platform                         Capcom        0        0     0.08   
34     Platform                         Capcom     0.29     0.22        0   
...         ...                            ...      ...      ...      ...   
16531  Platform                       Nintendo    14.59     7.06      4.7   
16563  Platform                            THQ     0.51      0.4        0   
16578  Platform  BMG Interactive Entertainment     320K     0.22        0   
16579  Platform                         Capcom     0.05     0.02        0   
16593  Platform                  Vivendi Games     0.15     0.04        0   

      Other_Sales  
4            0.04  
20           0.15  
25              0  
28              0  
34           0.08  
...           ...  
16531        2.26  
16563        0.11  
16578        0.04  
16579           0  
16593        0.01  

[884 rows x 10 columns]

###EDA-PUblisher

In [ ]:
df[df.Publisher.isnull()]

Unnamed: 0                                               Name Platform  \
272           273                                 UK Truck Simulator       PC   
1313         1314     All Grown Up!: Game Boy Advance Video Volume 1      GBA   
1560         1561  Teenage Mutant Ninja Turtles: Game Boy Advance...      GBA   
1789         1790  Nicktoons Collection: Game Boy Advance Video V...      GBA   
1879         1880                     Monster Hunter Frontier Online      PS3   
1961         1962  Nicktoons Collection: Game Boy Advance Video V...      GBA   
2070         2071                                     World of Tanks     X360   
2970         2971                               Chou Soujuu Mecha MG       DS   
3184         3185                       Action Man-Operation Extreme       PS   
3412         3413                                  Dance with Devils      PSV   
3533         3534                          B.L.U.E.: Legend of Water       PS   
3562         3563                         wwe Smackdown vs. Raw 2006      PS2   
3639         3640                    Homeworld Remastered Collection       PC   
3684         3685    Dragon Ball GT: Game Boy Advance Video Volume 1      GBA   
4308         4309  Umineko no Naku Koro ni San: Shinjitsu to Gens...      PS3   
4367         4368  Strawberry Shortcake: Game Boy Advance Video V...      GBA   
5207         5208  Cartoon Network Collection: Game Boy Advance V...      GBA   
5245         5246  Cartoon Network Collection: Game Boy Advance V...      GBA   
5387         5388  Moshi, Kono Sekai ni Kami-sama ga Iru to suru ...      PSV   
6209         6210             Atsumare! Power Pro Kun no DS Koushien       DS   
6392         6393                        Brothers in Arms: Furious 4     X360   
7370         7371                                       Stronghold 3       PC   
7984         7985                   Disney Channel Collection Vol. 1      GBA   
8038         8039  Super Robot Monkey Team: Game Boy Advance Vide...      GBA   
8236         8237                                 Bentley's Hackpack      GBA   
8355         8356  Cartoon Network Collection: Game Boy Advance V...      GBA   
8651         8652                                 Sonic the Hedgehog      PS3   
8873         8874                    Legends of Oz: Dorothy's Return      3DS   
9128         9129  The Fairly Odd Parents: Game Boy Advance Video...      GBA   
9684         9685      Dragon Ball Z: Budokai Tenkaichi 2 (JP sales)      Wii   
9877         9878                                       Mario Tennis      3DS   
10156       10157  The Legend of Zelda: The Minish Cap(weekly JP ...      GBA   
10223       10224                   Demolition Company: Gold Edition       PC   
11309       11310                                     Triple Play 99       PS   
11498       11499  Nicktoons Collection: Game Boy Advance Video V...      GBA   
11641       11642             AKB1/48: Idol to Guam de Koishitara...     X360   
11676       11677                             Farming Simulator 2011       PC   
12126       12127                                      Silverlicious       DS   
12176       12177  Dora the Explorer: Game Boy Advance Video Volu...      GBA   
12468       12469  The Fairly Odd Parents: Game Boy Advance Video...      GBA   
12560       12561  SpongeBob SquarePants: Game Boy Advance Video ...      GBA   
12900       12901  SpongeBob SquarePants: Game Boy Advance Video ...      GBA   
13182       13183           Sonic X: Game Boy Advance Video Volume 1      GBA   
13204       13205                                            Xia-Xia       DS   
13450       13451                                Bikkuriman Daijiten       DS   
13766       13767                                       Housekeeping       DS   
13810       13811  Pokémon: Johto Photo Finish: Game Boy Advance ...      GBA   
13837       13838  Cartoon Network Collection: Game Boy Advance V...      GBA   
13923       13924       Prinny: Can I 

In [ ]:
df.Publisher.unique()

array(['Destineer', 'Namco Bandai Games', 'Alchemist', 'Zoo Games',
       'D3Publisher', 'Konami Digital Entertainment', 'Deep Silver',
       'Tecmo Koei', 'Sega', 'THQ', 'Empire Interactive', 'Ubisoft',
       'Paradox Interactive', '505 Games', 'Tomy Corporation',
       'Eidos Interactive', 'Microsoft Game Studios',
       'Nippon Ichi Software', 'Kalypso Media', 'Electronic Arts',
       'Sony Computer Entertainment', 'Activision',
       'Tru Blu Entertainment', 'Warner Bros. Interactive Entertainment',
       'Midway Games', 'Mindscape', 'Wargaming.net', 'Capcom',
       'Acclaim Entertainment', 'Unknown', 'Nintendo',
       'Disney Interactive Studios', 'Crave Entertainment', 'Sunflowers',
       'Arc System Works', 'Milestone S.r.l', 'Kamui', 'Hudson Soft',
       'Square Enix', 'Focus Home Interactive', 'Take-Two Interactive',
       'Game Factory', 'Takara Tomy', 'SquareSoft', 'Global Star',
       'Infogrames', 'GT Interactive', 'Storm City Games',
       'Gathering of Dev

In [ ]:
df.Publisher.nunique()

578

###EDA-Sales

In [ ]:
df.sample(15)

Unnamed: 0                                   Name Platform    Year  \
10544       10545                               Catwoman       GC  2004.0   
11459       11460                                    Rio       DS  2011.0   
128           129              Rozen Maiden: Gebetgarten      PS2  2007.0   
4246         4247                               ArmA III       PC  2013.0   
11662       11663                       Dragoneer's Aria      PSP  2007.0   
1198         1199                     NCAA Football 2003       XB  2002.0   
15903       15904             Yu-Gi-Oh! 5D's Tag Force 6      PSP  2011.0   
13951       13952                     Top Gun: Hard Lock     X360  2012.0   
8769         8770              Street Fighter Collection      SAT  1996.0   
5863         5864                     Samurai Warriors 2      PS2  2006.0   
10609       10610  Disney's Story Studio: Disney's Mulan       PS  1999.0   
395           396       The Settlers II 10th Anniversary       PC  2006.0   
7373         7374          Enslaved: Odyssey to the West      PS3  2010.0   
8782         8783    Harry Potter and the Goblet of Fire       XB  2005.0   
7120         7121                Call of Duty: Black Ops      Wii  2010.0   

              Genre                     Publisher NA_Sales EU_Sales JP_Sales  \
10544        Action               Electronic Arts      50K     0.01        0   
11459          Misc                           THQ     0.12     0.03        0   
128        Fighting                         Taito        0       0K     0.01   
4246        Shooter           Bohemia Interactive        0      0.1        0   
11662  Role-Playing                    Tecmo Koei     0.02        0     0.05   
1198         Sports               Electronic Arts     0.39     0.11        0   
15903      Strategy  Konami Digital Entertainment        0        0     0.16   
13951        Action                     505 Games     0.03     0.02        0   
8769       Fighting                        Capcom        0        0     0.11   
5863         Action                    Tecmo Koei     0.02     0.02     0.57   
10609          Misc   Sony Computer Entertainment     0.09     0.06        0   
395          Action                       Ubisoft        0     0.02        0   
7373         Action            Namco Bandai Games     0.24     0.24     0.01   
8782         Action               Electronic Arts     0.18     0.05        0   
7120        Shooter                    Activision     0.82     0.47        0   

      Other_Sales  
10544           0  
11459        0.01  
128             0  
4246         0.02  
11662           0  
1198         0.02  
15903           0  
13951           0  
8769            0  
5863         0.01  
10609        0.01  
395             0  
7373         0.09  
8782         0.01  
7120         0.12

In [ ]:
# K, M 있는 것만 꺼내보자
# foo[foo.b.str.contains('oo', regex= True, na=False)]

# df.columns
# 'NA_Sales', 'EU_Sales', 'JP_Sales', 'Other_Sales'

df[df.NA_Sales.str.contains('M',regex=True, na=False)]

Unnamed: 0                                        Name Platform  \
447           448                  Medal of Honor: Warfighter     X360   
458           459                                     F1 2010      PS3   
577           578                               Mario Kart 64      N64   
625           626                                  The Hobbit      GBA   
672           673  Star Wars The Clone Wars: Lightsaber Duels      Wii   
...           ...                                         ...      ...   
15986       15987                 Juiced 2: Hot Import Nights       PC   
16152       16153                       Taiko no Tatsujin Wii      Wii   
16169       16170                    Medal of Honor: Airborne     X360   
16408       16409               NASCAR 06: Total Team Control      PS2   
16482       16483            Super Robot Taisen F Kanketsuhen      SAT   

         Year     Genre           Publisher NA_Sales EU_Sales JP_Sales  \
447    2012.0    Action     Electronic Arts    0.58M     0.47     0.01   
458    2010.0    Racing         Codemasters    0.25M     0.76     0.08   
577    1996.0    Racing            Nintendo    5.55M     1.94     2.23   
625    2003.0  Platform       Vivendi Games    0.14M     0.05        0   
672    2008.0    Action           LucasArts    1.23M      0.3        0   
...       ...       ...                 ...      ...      ...      ...   
15986  2007.0    Racing                 THQ       0M     0.01        0   
16152  2008.0      Misc  Namco Bandai Games       0M        0     0.67   
16169  2007.0   Shooter     Electronic Arts    0.45M     0.53     0.01   
16408  2005.0    Racing     Electronic Arts    0.53M     0.02        0   
16482    98.0  Strategy           Banpresto       0M        0     0.55   

      Other_Sales  
447           0.1  
458          0.27  
577          0.15  
625             0  
672          0.14  
...           ...  
15986           0  
16152           0  
16169        0.12  
16408        0.09  
16482           0  

[100 rows x 10 columns]

In [ ]:
df[df.NA_Sales.str.contains('K',regex=True, na=False)]

Unnamed: 0                                            Name Platform  \
10             11                                     Ford Racing       PS   
45             46                    Resident Evil: Revelations 2      PS3   
145           146                              TOCA Race Driver 3       PC   
878           879  Tears to Tiara Anecdotes: The Secret of Avalon      PS3   
1086         1087                       Pro Evolution Soccer 2014      3DS   
...           ...                                             ...      ...   
15647       15648                       Championship Manager 2010       PC   
16194       16195                            Vin Diesel: Wheelman      PS3   
16330       16331                         Midway Arcade Treasures      PS2   
16492       16493                                   Mortal Kombat      PSV   
16578       16579                            Gex: Enter the Gecko       PS   

         Year     Genre                               Publisher NA_Sales  \
10     2001.0    Racing                      Empire Interactive     480K   
45     2015.0    Action                                  Capcom      60K   
145    2006.0    Racing                             Codemasters       0K   
878    2009.0  Strategy                               Aqua Plus       0K   
1086   2013.0    Action            Konami Digital Entertainment       0K   
...       ...       ...                                     ...      ...   
15647  2009.0    Sports                       Eidos Interactive       0K   
16194  2009.0    Racing                                 Ubisoft     120K   
16330  2003.0      Misc                            Midway Games     720K   
16492  2012.0  Fighting  Warner Bros. Interactive Entertainment     470K   
16578  1998.0  Platform           BMG Interactive Entertainment     320K   

      EU_Sales JP_Sales Other_Sales  
10       0.33M       0K        0.06  
45        0.08     0.16        0.03  
145       0.02        0           0  
878          0     0.04           0  
1086         0     0.19           0  
...        ...      ...         ...  
15647     0.01        0           0  
16194     0.15        0        0.06  
16330     0.56        0        0.19  
16492     0.18        0        0.12  
16578     0.22        0        0.04  

[100 rows x 10 columns]

In [ ]:
df[df.iloc[:,-1].str.contains('M',regex=True, na=False)]

Unnamed: 0                                               Name Platform  \
64             65                                    SD Gundam Force      GBA   
73             74                       Teenage Mutant Ninja Turtles       XB   
209           210              Harry Potter and the Sorcerer's Stone       XB   
220           221          Dragon Quest IV: Michibikareshi Monotachi       PS   
243           244                 Seaman 2: Peking Genjin Ikusei Kit      PS2   
...           ...                                                ...      ...   
15998       15999                          Pro Yakyuu Family Stadium      NES   
16036       16037                                      Point Blank 2       PS   
16074       16075                     The Black Eyed Peas Experience      Wii   
16141       16142  World Championship Poker: Featuring Howard Led...      PS2   
16412       16413  Yu-Gi-Oh! Ultimate Masters: World Championship...      GBA   

         Year         Genre                     Publisher NA_Sales EU_Sales  \
64     2004.0       Shooter            Namco Bandai Games     0.01        0   
73     2003.0        Action  Konami Digital Entertainment     0.22     0.06   
209    2003.0        Action               Electronic Arts     0.05     0.02   
220    2001.0  Role-Playing              Enix Corporation        0        0   
243    2007.0    Simulation                          Sega        0        0   
...       ...           ...                           ...      ...      ...   
15998  1986.0        Sports            Namco Bandai Games     0.15        0   
16036  1998.0       Shooter            Namco Bandai Games     0.02     0.02   
16074  2011.0          Misc                       Ubisoft     0.39      0.2   
16141  2006.0          Misc            Oxygen Interactive     0.82     0.64   
16412  2006.0          Misc  Konami Digital Entertainment     0.14     0.05   

      JP_Sales Other_Sales  
64           0          0M  
73           0       0.01M  
209          0          0M  
220        1.2       0.01M  
243       0.06          0M  
...        ...         ...  
15998     2.05          0M  
16036        0          0M  
16074        0       0.06M  
16141        0       0.22M  
16412     0.07       0.01M  

[100 rows x 10 columns]

####Codes to test 🙅

In [ ]:
df_test = df.copy()

#####<font color = 'grey'>not working</font>

Delete non-number in characters using regular expression

`re.sub(pattern, new, string)`<br>
string에서 pattern에 해당하는 부분을 new로 대체하라는 의미 <br>
(replace 'pattern' with 'new' in 'string')

```python
# 함수를 만들어 구현
import re

def convert(phone_number):
  phone_number = re.sub(r"/D","", phone_number)[-8:]   #\D = [^0-9] \d = [0-9]
  final_number = '010' + '-' + phone_number[:4] + '-' + phone_number[-4:]
  return final_number
  
# 만든 함수를 .apply() 함수를 이용해 적용
patients_clean.전화번호 = patients_clean.전화번호.apply(convert)
```

[pandas.DataFrame.astype](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.astype.html)
- DataFrame.astype(dtype, copy=True, errors='raise')
- Cast a pandas object to a specified dtype dtype. <pr>

[Column에서 astype을 쓰는 방법](https://cosmosproject.tistory.com/123)
```python
df_test.loc[:, ['col1']] = df_test.loc[:, ['col1']].astype('str')
```

In [ ]:
import re

def convert(sales_figure):
  sales_figure = re.sub(r"\D","",sales_figure)
  # sales_figure = sales_figure.astype(float)
  return sales_figure

[range(start, stop, step)](https://www.geeksforgeeks.org/python-range-function/)

- start: integer starting from which the sequence of integers is to be returned
- stop: integer before which the sequence of integers is to be returned. The range of integers ends at stop – 1.
- step: integer value which determines the increment between each integer in the sequence

In [ ]:
for i in range(1,5,1):
  df_test.iloc[:,-i] = df_test.iloc[:,-i].apply(convert)

AttributeError: ignored

In [ ]:
for i in range(1,5,1):
  df_test.iloc[:,-i] = df_test.iloc[:,-i].astype(float)

ValueError: ignored

#####**working**

In [ ]:
import re

def convert(sales_figure):
  sales_figure = re.sub(r"\D","",sales_figure)

  return sales_figure

In [ ]:
df_test.iloc[:,-1]=df_test.iloc[:,-1].apply(convert)

In [ ]:
df_test.iloc[[64]]

Unnamed: 0             Name Platform    Year    Genre           Publisher  \
64          65  SD Gundam Force      GBA  2004.0  Shooter  Namco Bandai Games   

   NA_Sales EU_Sales JP_Sales Other_Sales  
64     0.01        0        0          0M

In [ ]:
for i in range(1,5,1):
  df_test.iloc[:,-i] = df_test.iloc[:,-i].apply(convert)

#####<font color = 'grey'>not working</font>

In [ ]:
index_list = df[df.NA_Sales.str.contains('M',regex=True, na=False) | df.EU_Sales.str.contains('M',regex=True, na=False) | \
                df.JP_Sales.str.contains('M',regex=True, na=False) | df.Other_Sales.str.contains('M',regex=True, na=False) ].index.tolist()
index_list

In [ ]:
df_test[[index_list,:]]

SyntaxError: ignored

#####working

**First of all, `search` for the rows to be fixed. This process results in a list of True/False**

```python
import re

def has_errors(inputString):
  return bool(re.search('[_,!,.,\d+]', inputString))
  
  # (설명)re.search --- inputString arguments에 _ ! . \d+ 숫자(한 개 이상 +) 찾으면 
  # (설명) bool --- True/False로 나온다
```

```python
# 이름에 섞여 있는 오류 데이터를 확인
error_condition = patients_clean.이름.apply(has_errors)
error_names = patients_clean[error_condition]
error_names

 # (설명) error_condition에 어떤 조건을 넣고, 그 조건이 맞는 데이터프레임을 error_name으로 불러옴
 ```

 
**Second, `replace` characters in values of the rows that has been searched in the first step**
```python
# 코드가 정확하게 오류를 제거하는지 확인
error_names.이름 = error_names.이름.str.replace('[_,!,.,\d+]','')
error_names

# (설명) str.replace()는 시리즈에서 _,!,.,\d+를 없애는 것으로 대체함
```
**Finally, apply the same `str.replace` method to the target DataFrame**
```python
# 제거된 것을 확인하였으면, 환자데이터 이름 컬럼에 적용
patients_clean.이름 = patients_clean.이름.str.replace('[_,!,.,\d+]','')
```

In [ ]:
import re

def has_errors(inputString):
  return bool(re.search('[M,K]',inputString))

In [ ]:
# call list of boolean that indicates number(False) or non-number(True)
not_num_list_NA = df_test.NA_Sales.apply(has_errors)
not_num_list_EU = df_test.EU_Sales.apply(has_errors)
not_num_list_JP = df_test.JP_Sales.apply(has_errors)
not_num_list_Other = df_test.Other_Sales.apply(has_errors)

In [ ]:
df_test_figure_err = df_test[not_num_list_NA | not_num_list_EU | not_num_list_JP | not_num_list_Other]
df_test_figure_err

Unnamed: 0                                   Name Platform    Year  \
10             11                            Ford Racing       PS  2001.0   
36             37              Advance Wars: Dual Strike       DS  2005.0   
45             46           Resident Evil: Revelations 2      PS3  2015.0   
64             65                        SD Gundam Force      GBA  2004.0   
73             74           Teenage Mutant Ninja Turtles       XB  2003.0   
...           ...                                    ...      ...     ...   
16525       16526                        Planet Monsters      GBA  2001.0   
16571       16572                           Koihime Enbu      PS4  2016.0   
16578       16579                   Gex: Enter the Gecko       PS  1998.0   
16580       16581  Captain Morgane and the Golden Turtle       DS  2012.0   
16591       16592                        Madden NFL 2005      PS2  2004.0   

           Genre                      Publisher NA_Sales EU_Sales JP_Sales  \
10        Racing             Empire Interactive     480K    0.33M       0K   
36      Strategy                       Nintendo      0.3    0.03M     0.04   
45        Action                         Capcom      60K     0.08     0.16   
64       Shooter             Namco Bandai Games     0.01        0        0   
73        Action   Konami Digital Entertainment     0.22     0.06        0   
...          ...                            ...      ...      ...      ...   
16525     Action                          Titus     0.01       0K        0   
16571   Fighting                           Yeti        0       0M     0.02   
16578   Platform  BMG Interactive Entertainment     320K     0.22        0   
16580  Adventure             Reef Entertainment        0     0.03       0M   
16591     Sports                Electronic Arts     4.18     0.26     0.01   

      Other_Sales  
10           0.06  
36           0.03  
45           0.03  
64             0M  
73          0.01M  
...           ...  
16525           0  
16571           0  
16578        0.04  
16580           0  
16591         80K  

[785 rows x 10 columns]

#####<font color = 'grey'>not working</font>

In [ ]:
test_figure_err_boolean = [not_num_list_NA | not_num_list_EU | not_num_list_JP | not_num_list_Other]
test_figure_err_boolean

[0        False
 1        False
 2        False
 3        False
 4        False
          ...  
 16593    False
 16594    False
 16595    False
 16596    False
 16597    False
 Length: 16598, dtype: bool]

In [ ]:
df_test[test_figure_err_boolean]

KeyError: ignored

#####**working**

In [ ]:
# df_test_figure_err.iloc[:,-4:-1] = df_test_figure_err.iloc[:,-4:-1].str.replace("[M,K]","")
# DataFrame에선 str 안 돌아가니까, column단위에서 보기 위해서 for문 돌림

for i in range(1,5,1):
  df_test_figure_err.iloc[:,-i] = df_test_figure_err.iloc[:,-i].str.replace("[M,K]","")

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:5: FutureWarning: The default value of regex will change from True to False in a future version.
  """
/usr/local/lib/python3.7/dist-packages/pandas/core/indexing.py:1773: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(ilocs[0], value, pi)


[iloc[ ] to choose the last column](https://stackoverflow.com/questions/37512079/python-pandas-why-does-df-iloc-1-values-for-my-training-data-select-till)

Python list slicing syntax states that for a:b it will get a and everything upto but not including b. a: will get a and everything after it. :b will get everything before b but not b. The list index of -1 refers to the last element. :-1 adheres to the same standards as above in that this gets everything before the last element but not the last element. If you want the last element included use :.

In [ ]:
df_test_figure_err.iloc[:,-4:] #마지막 column이 나오려면 -1을 쓰면 그 전까지 나오니까, 아무것도 안 쓰면 된다

NA_Sales EU_Sales JP_Sales Other_Sales
10         480     0.33        0        0.06
36         0.3     0.03     0.04        0.03
45          60     0.08     0.16        0.03
64        0.01        0        0           0
73        0.22     0.06        0        0.01
...        ...      ...      ...         ...
16525     0.01        0        0           0
16571        0        0     0.02           0
16578      320     0.22        0        0.04
16580        0     0.03        0           0
16591     4.18     0.26     0.01          80

[785 rows x 4 columns]

In [ ]:
df_test.sample(5)

Unnamed: 0                                               Name Platform  \
3769         3770                    Barbie and the Three Musketeers       DS   
10336       10337                       Harukanaru Toki no Naka de 6      PSV   
11749       11750                                Tom Clancy's EndWar      PS3   
4795         4796  Tim Burton's The Nightmare Before Christmas: O...       XB   
12722       12723                                     Rayman Origins      3DS   

         Year      Genre                   Publisher NA_Sales EU_Sales  \
3769   2009.0  Adventure                  Activision     0.12        0   
10336  2015.0     Action                  Tecmo Koei        0        0   
11749  2008.0   Strategy                     Ubisoft     0.32     0.22   
4795   2005.0  Adventure  Disney Interactive Studios     0.06     0.02   
12722  2012.0   Platform                     Ubisoft     0.06     0.08   

      JP_Sales Other_Sales  
3769         0        0.01  
10336     0.03           0  
11749     0.01         0.1  
4795         0           0  
12722        0        0.01

In [ ]:
import re

def has_errors(inputString):
  return bool(re.search('K',inputString))


for i in range(len(df_test.shape[0])):
  if df_test.NA_Sales[i].apply(has_errors) == True:
    df_test.NA_Sales[i] = df_test.NA_Sales[i].str.replace("K","")
    df_test.NA_Sales[i] = df_test.NA_Sales[i].astype({}'int')


# import re

# def has_errors(inputString):
#   return bool(re.search('K',inputString))

# df_test[df_test.NA_Sales.apply(has_errors) == True]

TypeError: ignored

####Codes to proceed 🙆

In [ ]:
df_clean = df.copy()

In [ ]:
df_clean.sample(5)

Unnamed: 0                               Name Platform    Year  \
10149       10150       Midnight Club 3: DUB Edition       XB  2005.0   
2891         2892               Zangeki no Reginleiv      Wii  2010.0   
843           844                 MX vs. ATV Untamed       DS  2007.0   
10253       10254  Fatal Frame II: Crimson Butterfly      PS2  2003.0   
11028       11029     Ready 2 Rumble Boxing: Round 2       PS  2000.0   

          Genre             Publisher NA_Sales EU_Sales JP_Sales Other_Sales  
10149    Racing  Take-Two Interactive     0.61      0.1        0        0.03  
2891     Action              Nintendo        0        0     0.11           0  
843      Racing                   THQ     0.11        0        0        0.01  
10253    Action               Ubisoft     0.08     0.06        0        0.02  
11028  Fighting          Midway Games     0.29      0.2        0        0.03

*Do not drop K, M at the same time because K, M are hinting unit* </br>
*Instead, drop K and multiply 0.001 at the same time and just drop M*

In [ ]:
# for i in range(1,5,1):
#   df_clean.iloc[:,-i] = df_clean.iloc[:,-i].str.replace("M","")

# for i in range(1,5,1):
#   df_clean.iloc[:,-i] = df_clean.iloc[:,-i].str.

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: FutureWarning: The default value of regex will change from True to False in a future version.
  


In [ ]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16598 entries, 0 to 16597
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Unnamed: 0   16598 non-null  int64  
 1   Name         16598 non-null  object 
 2   Platform     16598 non-null  object 
 3   Year         16327 non-null  float64
 4   Genre        16548 non-null  object 
 5   Publisher    16540 non-null  object 
 6   NA_Sales     16598 non-null  object 
 7   EU_Sales     16598 non-null  object 
 8   JP_Sales     16598 non-null  object 
 9   Other_Sales  16598 non-null  object 
dtypes: float64(1), int64(1), object(8)
memory usage: 1.3+ MB


###EDA-Duplicated

In [ ]:
df[df.duplicated()]

Empty DataFrame
Columns: [Unnamed: 0, Name, Platform, Year, Genre, Publisher, NA_Sales, EU_Sales, JP_Sales, Other_Sales]
Index: []

###EDA result

Quality Issue
- Null: year 271, Genre 50, Publisher 58 🟢
- Year data type : float to integer, drop 271 null rows (1.6% of raw data) 🟢
- Sales : K,M, unit alignment ex. 0M, 50K 🟢
- Genre : drop 50 null rows, dtype to category 🟢
- Publisher : drop 58 null rows 🟢
- Unnamed:0  : primary key to index 🟢

Structure Issue
- Sales : melt into one column 🟢


##Wrangling

In [4]:
df_clean = df.copy()

####Wrangling-Genre

[pandas.DataFrame.dropna](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.dropna.html)
- DataFrame.dropna(axis=0, how='any', thresh=None, subset=None, inplace=False)
  - subset: Labels along other axis to consider, e.g. if you are dropping rows these would be a list of columns to include.

In [5]:
# Drop rows that are NaN in Genre
df_clean = df_clean.dropna(subset=['Genre'])

In [6]:
df_clean.shape[0]

16548

In [7]:
# Transform data type of Genre into categorical type
df_clean.Genre = df_clean.Genre.astype('category')

/usr/local/lib/python3.7/dist-packages/pandas/core/generic.py:5516: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


In [8]:
df_clean.Genre.dtype

CategoricalDtype(categories=['Action', 'Adventure', 'Fighting', 'Misc', 'Platform',
                  'Puzzle', 'Racing', 'Role-Playing', 'Shooter', 'Simulation',
                  'Sports', 'Strategy'],
, ordered=False)

###Wrangling-Year

In [9]:
df_clean.head()

Unnamed: 0                                            Name Platform  \
0           1                    Candace Kane's Candy Factory       DS   
1           2                                  The Munchables      Wii   
2           3  Otome wa Oanesama Boku ni Koi Shiteru Portable      PSP   
3           4                Deal or No Deal: Special Edition       DS   
4           5       Ben 10 Ultimate Alien: Cosmic Destruction      PS3   

     Year      Genre           Publisher NA_Sales EU_Sales JP_Sales  \
0  2008.0     Action           Destineer     0.04        0        0   
1  2009.0     Action  Namco Bandai Games     0.17        0        0   
2  2010.0  Adventure           Alchemist        0        0     0.02   
3  2010.0       Misc           Zoo Games     0.04        0        0   
4  2010.0   Platform         D3Publisher     0.12     0.09        0   

  Other_Sales  
0           0  
1        0.01  
2           0  
3           0  
4        0.04

In [10]:
# Drop NaN in Year
df_clean = df_clean.dropna(subset=['Year'])

In [11]:
df_clean.shape

(16277, 10)

In [12]:
# Change data type of Year into integer
df_clean.Year = df_clean.Year.astype('int')

/usr/local/lib/python3.7/dist-packages/pandas/core/generic.py:5516: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


In [13]:
df_clean.Year.dtype

dtype('int64')

###Wrangling-Publisher


In [14]:
df_clean.shape

(16277, 10)

In [15]:
df_clean = df_clean.dropna(subset=['Publisher'])
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16241 entries, 0 to 16597
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   Unnamed: 0   16241 non-null  int64   
 1   Name         16241 non-null  object  
 2   Platform     16241 non-null  object  
 3   Year         16241 non-null  int64   
 4   Genre        16241 non-null  category
 5   Publisher    16241 non-null  object  
 6   NA_Sales     16241 non-null  object  
 7   EU_Sales     16241 non-null  object  
 8   JP_Sales     16241 non-null  object  
 9   Other_Sales  16241 non-null  object  
dtypes: category(1), int64(2), object(7)
memory usage: 1.3+ MB


###Wrangling-Sales

In [16]:
sales_list = list(df_clean)[-4:]

In [17]:
df_clean[sales_list].sample(10)

NA_Sales EU_Sales JP_Sales Other_Sales
5680         0     0.04        0           0
6732       0.1     0.08     0.07        0.04
13402     0.06     0.02        0           0
12293        0     0.19        0        0.05
12265        0     0.02        0           0
10283      0.1     0.01        0        0.01
1373      0.76     0.36     0.05        0.18
210       0.29     0.02        0           0
16238     0.05      0.1        0        0.03
10341     0.12     0.03        0        0.01

In [18]:
import re

def has_units(inputString):
  return bool(re.search('[M,K]', inputString))

In [19]:
# M, K를 포함하는 rows가 True인 boolean 리스트
Unit_NA = df_clean.NA_Sales.apply(has_units)
Unit_EU = df_clean.EU_Sales.apply(has_units)
Unit_JP = df_clean.JP_Sales.apply(has_units)
Unit_Other = df_clean.Other_Sales.apply(has_units)

In [20]:
df_clean[Unit_NA | Unit_EU | Unit_JP | Unit_Other].query(' Year == 2011 ').sample(20)

Unnamed: 0                                           Name Platform  \
10274       10275     Phineas and Ferb: Across the 2nd Dimension      PS3   
7505         7506                               Pac-Man Party 3D      3DS   
5561         5562                  Meikyuu Cross Blood: Reloaded     X360   
14250       14251                              Airport Simulator       PC   
15270       15271        Tales of the World: Radiant Mythology 3      PSP   
8421         8422                                   Grease Dance      PS3   
13487       13488                         Dance! It's Your Stage      PS3   
493           494                             Rise of Nightmares     X360   
2948         2949                             Army Corps of Hell      PSV   
2495         2496                                    Bulletstorm     X360   
16347       16348                            Aladdin Magic Racer      Wii   
13168       13169                                 Rayman Origins     X360   
8632         8633                                   MotoGP 10/11     X360   
15706       15707                         Shadow of the Colossus      PS3   
6112         6113                              Jewel Time Deluxe       DS   
15288       15289                                     Dark Souls     X360   
1975         1976                                        FIFA 12      3DS   
14707       14708                                     Fit in Six      Wii   
9091         9092  Tom Clancy's Splinter Cell Classic Trilogy HD      PS3   
8827         8828                                 Arcana Heart 3     X360   

       Year         Genre                    Publisher NA_Sales EU_Sales  \
10274  2011        Action   Disney Interactive Studios     0.15    0.12M   
7505   2011          Misc           Namco Bandai Games     0.09      40K   
5561   2011        Action                      Unknown        0        0   
14250  2011    Simulation                      Unknown        0     0.02   
15270  2011  Role-Playing           Namco Bandai Games        0        0   
8421   2011          Misc                    505 Games     0.05      20K   
13487  2011          Misc            DTP Entertainment        0     0.02   
493    2011        Action                         Sega     0.28     0.13   
2948   2011        Action                  Square Enix     0.05    0.04M   
2495   2011       Shooter              Electronic Arts    0.61M    0.32M   
16347  2011        Racing          Big Ben Interactive     0.01       0M   
13168  2011      Platform                      Ubisoft     0.38     0.34   
8632   2011        Racing                       Capcom       0M     0.05   
15706  2011        Action  Sony Computer Entertainment        0       0M   
6112   2011        Puzzle                      O-Games     0.04      30K   
15288  2011  Role-Playing           Namco Bandai Games     0.64     0.28   
1975   2011        Sports              Electronic Arts     0.08     0.26   
14707  2011        Sports                      Ubisoft     0.11        0   
9091   2011        Action                      Ubisoft     0.02        0   
8827   2011      Fighting                        PQube        0    0.03M   

      JP_Sales Other_Sales  
10274        0        0.05  
7505         0        0.01  
5561      0.02          0K  
14250        0          0M  
15270     0.28          0M  
8421         0        0.01  
13487       0M        0.01  
493      0.01M        0.04  
2948         0        0.02  
2495      0.01        0.08  
16347        0           0  
13168        0       0.08M  
8632         0        0.01  
15706     0.04           0  
6112         0        0.01  
15288        0       0.08M  
1975         0       0.05M  
14707       0K        0.01  
9091         0          0K  
8827       20K           0

In [21]:
df_clean.loc[[8010, 6929]]

Unnamed: 0           Name Platform  Year    Genre             Publisher  \
8010        8011  Battlefield 3     X360  2011  Shooter       Electronic Arts   
6929        6930       NBA 2K12      PS3  2011   Sports  Take-Two Interactive   

     NA_Sales EU_Sales JP_Sales Other_Sales  
8010     4.46    2130K     0.06        0.69  
6929    1610K     0.27     0.05        0.18

**(결론) 출시한 해(1년)에 팔린 장수로 보자**
> Year : 출시년도</br>
> K, M 안 붙어있는 수치 : million </br>
> [Battlefield 3 copies sold till Dec 2011](https://www.eurogamer.net/battlefield-3-sales-hit-8-million) </br>
Battlefield 3 has now sold eight million copies since launch, publisher EA has revealed.</br>
Data -> 7.34M</br>
> [NBA 2K12 copies sold till Feb 2012](https://www.gamespot.com/articles/red-dead-redemption-sells-13m-nba-2k12-hits-4m/1100-6349683/) </br>
NBA 2K12 sales have reached 4 million units </br>
Data -> 2,110K

In [22]:
df_clean_MK = df_clean[Unit_NA | Unit_EU | Unit_JP | Unit_Other]

In [23]:
# M은 떼어내기

df_clean.NA_Sales = df_clean.NA_Sales.str.replace('M',"")
df_clean.EU_Sales = df_clean.EU_Sales.str.replace('M',"")
df_clean.JP_Sales = df_clean.JP_Sales.str.replace('M',"")
df_clean.Other_Sales = df_clean.Other_Sales.str.replace('M',"")


In [24]:
df_clean[df_clean.index.isin(df_clean_MK.index)].sample(10)

Unnamed: 0                                               Name Platform  \
9472         9473                                      Darksiders II     XOne   
642           643            LEGO Star Wars II: The Original Trilogy       GC   
15385       15386                         Thrillville: Off the Rails     X360   
15808       15809                  Final Fantasy XIV: A Realm Reborn       PC   
8992         8993  Virtua Fighter CG Portrait Series Vol.1: Sarah...      SAT   
16482       16483                   Super Robot Taisen F Kanketsuhen      SAT   
1650         1651                        Test Drive: Ferrari Legends      PS3   
8010         8011                                      Battlefield 3     X360   
12158       12159                              Front Mission Evolved     X360   
10389       10390                                       Lost Horizon       PC   

       Year         Genre        Publisher NA_Sales EU_Sales JP_Sales  \
9472   2015        Action     Nordic Games     0.03     0.02        0   
642    2006        Action        LucasArts     0.75     0.19        0   
15385  2007      Strategy        LucasArts     0.13     0.02        0   
15808  2010  Role-Playing      Square Enix     0.86     0.48       0K   
8992   1995          Misc             Sega        0        0     0.11   
16482    98      Strategy        Banpresto        0        0     0.55   
1650   2012        Racing          Unknown     0.04        0        0   
8010   2011       Shooter  Electronic Arts     4.46    2130K     0.06   
12158  2010       Shooter      Square Enix     100K     0.07     0.02   
10389  2010     Adventure      Deep Silver        0      0.1       0K   

      Other_Sales  
9472         0.01  
642           30K  
15385        0.01  
15808        0.17  
8992            0  
16482           0  
1650           0K  
8010         0.69  
12158        0.02  
10389        0.02

In [25]:
import re
import math

for i, sales in enumerate(df_clean.NA_Sales):
  if bool(re.search('K', sales)):
    df_clean.iloc[i,6] = re.sub(r"\D","",df_clean.iloc[i,6])
    df_clean.iloc[i,6] = float(df_clean.iloc[i,6])
    df_clean.iloc[i,6] = df_clean.iloc[i,6]*0.001

for i, sales in enumerate(df_clean.EU_Sales):
  if bool(re.search('K', sales)):
    df_clean.iloc[i,7] = re.sub(r"\D","",df_clean.iloc[i,7])
    df_clean.iloc[i,7] = float(df_clean.iloc[i,7])
    df_clean.iloc[i,7] = df_clean.iloc[i,7]*0.001

for i, sales in enumerate(df_clean.JP_Sales):
  if bool(re.search('K', sales)):
    df_clean.iloc[i,8] = re.sub(r"\D","",df_clean.iloc[i,8])
    df_clean.iloc[i,8]= float(df_clean.iloc[i,8])
    df_clean.iloc[i,8] = df_clean.iloc[i,8]*0.001

for i, sales in enumerate(df_clean.Other_Sales):
  if bool(re.search('K', sales)):
    df_clean.iloc[i,9] = re.sub(r"\D","",df_clean.iloc[i,9])
    df_clean.iloc[i,9] = float(df_clean.iloc[i,9])
    df_clean.iloc[i,9] = df_clean.iloc[i,9]*0.001

# .extract, .astype 쓸 때, 앞에 iloc[i,6] 다음에 붙으면 str 특징이 없어서? 안 돌아간다. 왜 그러냐.


In [26]:
df_clean[df_clean.index.isin(df_clean_MK.index)].sample(10)

Unnamed: 0                                 Name Platform  Year  \
13874       13875                         Classic Road       PS  1995   
10342       10343                              NBA 2K6       XB  2005   
13341       13342                          NASCAR Heat       PS  2000   
8276         8277   Olympic Summer Games: Atlanta 1996       PS  1996   
1895         1896               Heart no Kuni no Alice      PSP  2009   
8589         8590  The Walking Dead: Survival Instinct     WiiU  2013   
9244         9245                         Afro Samurai      PS3  2009   
5583         5584                      Digimon World 2       PS  2000   
3768         3769                              DaGeDar       DS  2011   
16157       16158                        The Suffering      PS2  2004   

              Genre               Publisher NA_Sales EU_Sales JP_Sales  \
13874        Sports      Victor Interactive        0        0     0.07   
10342        Action    Take-Two Interactive     0.27     0.08        0   
13341        Racing      Hasbro Interactive     0.22     0.15        0   
8276         Sports               U.S. Gold     0.06     0.04        0   
1895      Adventure               Prototype        0        0     0.01   
8589        Shooter              Activision     0.08     0.03        0   
9244         Action                   Atari     0.12     0.05        0   
5583   Role-Playing      Namco Bandai Games     0.23     0.16     0.16   
3768         Action  GameMill Entertainment     0.09        0        0   
16157        Action            Midway Games      0.2     0.16        0   

      Other_Sales  
13874           0  
10342        0.01  
13341        0.03  
8276         0.01  
1895            0  
8589         0.01  
9244         0.03  
5583         0.04  
3768         0.01  
16157        0.05

In [27]:
df_clean.loc[[8827]]

Unnamed: 0            Name Platform  Year     Genre Publisher NA_Sales  \
8827        8828  Arcana Heart 3     X360  2011  Fighting     PQube        0   

     EU_Sales JP_Sales Other_Sales  
8827     0.03     0.02           0

###Wrangling-Unnamed :0

In [28]:
df_clean.sample(10)

Unnamed: 0                                          Name Platform  \
11848       11849         Busou Renkin: Youkoso Papillon Park e      PS2   
6030         6031                   Chevrolet Camaro: Wild Ride       DS   
5376         5377                  Star Wars: Republic Commando       XB   
2389         2390  Call of Duty: Modern Warfare: Reflex Edition      Wii   
8309         8310           WRC 5: FIA World Rally Championship     XOne   
10442       10443                         From Russia With Love      PS2   
3789         3790            Ed, Edd n Eddy: The Mis-Edventures       GC   
6066         6067                  Dexter's Lab Chess Challenge      GBA   
1273         1274                             Star Fox: Assault       GC   
2971         2972                            Namco Museum Vol.3       PS   

       Year     Genre                    Publisher NA_Sales EU_Sales JP_Sales  \
11848  2007    Action        Marvelous Interactive        0        0     0.02   
6030   2010    Racing             Storm City Games     0.02        0        0   
5376   2005   Shooter                   Activision     0.48     0.17        0   
2389   2009   Shooter                   Activision     0.95     0.43        0   
8309   2015    Sports           Bigben Interactive        0     0.01        0   
10442  2005    Action              Electronic Arts     0.34     0.01        0   
3789   2005  Platform                 Midway Games     0.19     0.05        0   
6066   2002      Misc           BAM! Entertainment     0.14     0.05        0   
1273   2005   Shooter                     Nintendo     0.68     0.18     0.19   
2971   1996      Misc  Sony Computer Entertainment     2.28     1.55     0.16   

      Other_Sales  
11848           0  
6030            0  
5376         0.02  
2389         0.14  
8309            0  
10442        0.06  
3789         0.01  
6066            0  
1273         0.03  
2971         0.06

In [29]:
df_clean.tail()

Unnamed: 0                         Name Platform  Year     Genre  \
16593       16594      Ice Age 2: The Meltdown       GC  2006  Platform   
16594       16595  Rainbow Islands: Revolution      PSP  2005    Action   
16595       16596                     NBA 2K16      PS3  2015    Sports   
16596       16597  Toukiden: The Age of Demons      PSV  2013    Action   
16597       16598     The King of Fighters '95       PS  1996  Fighting   

                         Publisher NA_Sales EU_Sales JP_Sales Other_Sales  
16593                Vivendi Games     0.15     0.04        0        0.01  
16594            Rising Star Games     0.01        0        0           0  
16595         Take-Two Interactive     0.44     0.19     0.03        0.13  
16596                   Tecmo Koei     0.05     0.05     0.25        0.03  
16597  Sony Computer Entertainment        0        0     0.16        0.01

In [30]:
df_clean.shape

(16241, 10)

[pandas.DataFrame.set_index](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.set_index.html)
- Set the DataFrame index using existing columns.
- DataFrame.set_index(keys, drop=True, append=False, inplace=False, verify_integrity=False)
  ```python
  df.set_index('month')
        year  sale
  month
  1      2012    55
  4      2014    40
  7      2013    84
  10     2014    31
  ```

In [31]:
df_clean.set_index('Unnamed: 0', inplace=True)
df_clean

Name Platform  Year  \
Unnamed: 0                                                                  
1                             Candace Kane's Candy Factory       DS  2008   
2                                           The Munchables      Wii  2009   
3           Otome wa Oanesama Boku ni Koi Shiteru Portable      PSP  2010   
4                         Deal or No Deal: Special Edition       DS  2010   
5                Ben 10 Ultimate Alien: Cosmic Destruction      PS3  2010   
...                                                    ...      ...   ...   
16594                              Ice Age 2: The Meltdown       GC  2006   
16595                          Rainbow Islands: Revolution      PSP  2005   
16596                                             NBA 2K16      PS3  2015   
16597                          Toukiden: The Age of Demons      PSV  2013   
16598                             The King of Fighters '95       PS  1996   

                Genre                    Publisher NA_Sales EU_Sales JP_Sales  \
Unnamed: 0                                                                      
1              Action                    Destineer     0.04        0        0   
2              Action           Namco Bandai Games     0.17        0        0   
3           Adventure                    Alchemist        0        0     0.02   
4                Misc                    Zoo Games     0.04        0        0   
5            Platform                  D3Publisher     0.12     0.09        0   
...               ...                          ...      ...      ...      ...   
16594        Platform                Vivendi Games     0.15     0.04        0   
16595          Action            Rising Star Games     0.01        0        0   
16596          Sports         Take-Two Interactive     0.44     0.19     0.03   
16597          Action                   Tecmo Koei     0.05     0.05     0.25   
16598        Fighting  Sony Computer Entertainment        0        0     0.16   

           Other_Sales  
Unnamed: 0              
1                    0  
2                 0.01  
3                    0  
4                    0  
5                 0.04  
...                ...  
16594             0.01  
16595                0  
16596             0.13  
16597             0.03  
16598             0.01  

[16241 rows x 9 columns]

[How to Rename Columns in Pandas (With Examples)](https://www.statology.org/pandas-rename-columns/)
- Rename Specific Columns
```python
df.rename(columns = {'old_col1':'new_col1', 'old_col2':'new_col2'}, inplace = True)
```

[How to rename the index of a Pandas DataFrame in Python](https://www.adamsmith.haus/python/answers/how-to-rename-the-index-of-a-pandas-dataframe-in-python)
  ```python
  df = pd.DataFrame([[1, 2], [3, 4]], columns = list("ab"))
  df = df.set_index("a")

  df.index.names = ["index"]
  ```

In [32]:
df_clean.index.names = ["Game_Key"]

In [33]:
df_clean.head()

Name Platform  Year  \
Game_Key                                                                  
1                           Candace Kane's Candy Factory       DS  2008   
2                                         The Munchables      Wii  2009   
3         Otome wa Oanesama Boku ni Koi Shiteru Portable      PSP  2010   
4                       Deal or No Deal: Special Edition       DS  2010   
5              Ben 10 Ultimate Alien: Cosmic Destruction      PS3  2010   

              Genre           Publisher NA_Sales EU_Sales JP_Sales Other_Sales  
Game_Key                                                                        
1            Action           Destineer     0.04        0        0           0  
2            Action  Namco Bandai Games     0.17        0        0        0.01  
3         Adventure           Alchemist        0        0     0.02           0  
4              Misc           Zoo Games     0.04        0        0           0  
5          Platform         D3Publisher     0.12     0.09        0        0.04

###Wrangling-Sales structure

In [34]:
id_vars = list(df_clean)[0:5]
id_vars

['Name', 'Platform', 'Year', 'Genre', 'Publisher']

In [35]:
df_clean = pd.melt(df_clean, id_vars=id_vars, var_name='Region', value_name='Sales')

In [36]:
df_clean

Name Platform  Year  \
0                        Candace Kane's Candy Factory       DS  2008   
1                                      The Munchables      Wii  2009   
2      Otome wa Oanesama Boku ni Koi Shiteru Portable      PSP  2010   
3                    Deal or No Deal: Special Edition       DS  2010   
4           Ben 10 Ultimate Alien: Cosmic Destruction      PS3  2010   
...                                               ...      ...   ...   
64959                         Ice Age 2: The Meltdown       GC  2006   
64960                     Rainbow Islands: Revolution      PSP  2005   
64961                                        NBA 2K16      PS3  2015   
64962                     Toukiden: The Age of Demons      PSV  2013   
64963                        The King of Fighters '95       PS  1996   

           Genre                    Publisher       Region Sales  
0         Action                    Destineer     NA_Sales  0.04  
1         Action           Namco Bandai Games     NA_Sales  0.17  
2      Adventure                    Alchemist     NA_Sales     0  
3           Misc                    Zoo Games     NA_Sales  0.04  
4       Platform                  D3Publisher     NA_Sales  0.12  
...          ...                          ...          ...   ...  
64959   Platform                Vivendi Games  Other_Sales  0.01  
64960     Action            Rising Star Games  Other_Sales     0  
64961     Sports         Take-Two Interactive  Other_Sales  0.13  
64962     Action                   Tecmo Koei  Other_Sales  0.03  
64963   Fighting  Sony Computer Entertainment  Other_Sales  0.01  

[64964 rows x 7 columns]

In [37]:
# Region 이름 정리
df_clean.Region = df_clean.Region.str.replace('_Sales','')
df_clean

Name Platform  Year  \
0                        Candace Kane's Candy Factory       DS  2008   
1                                      The Munchables      Wii  2009   
2      Otome wa Oanesama Boku ni Koi Shiteru Portable      PSP  2010   
3                    Deal or No Deal: Special Edition       DS  2010   
4           Ben 10 Ultimate Alien: Cosmic Destruction      PS3  2010   
...                                               ...      ...   ...   
64959                         Ice Age 2: The Meltdown       GC  2006   
64960                     Rainbow Islands: Revolution      PSP  2005   
64961                                        NBA 2K16      PS3  2015   
64962                     Toukiden: The Age of Demons      PSV  2013   
64963                        The King of Fighters '95       PS  1996   

           Genre                    Publisher Region Sales  
0         Action                    Destineer     NA  0.04  
1         Action           Namco Bandai Games     NA  0.17  
2      Adventure                    Alchemist     NA     0  
3           Misc                    Zoo Games     NA  0.04  
4       Platform                  D3Publisher     NA  0.12  
...          ...                          ...    ...   ...  
64959   Platform                Vivendi Games  Other  0.01  
64960     Action            Rising Star Games  Other     0  
64961     Sports         Take-Two Interactive  Other  0.13  
64962     Action                   Tecmo Koei  Other  0.03  
64963   Fighting  Sony Computer Entertainment  Other  0.01  

[64964 rows x 7 columns]

###Wrangling-Sales Data type

In [38]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 64964 entries, 0 to 64963
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype   
---  ------     --------------  -----   
 0   Name       64964 non-null  object  
 1   Platform   64964 non-null  object  
 2   Year       64964 non-null  int64   
 3   Genre      64964 non-null  category
 4   Publisher  64964 non-null  object  
 5   Region     64964 non-null  object  
 6   Sales      64964 non-null  object  
dtypes: category(1), int64(1), object(5)
memory usage: 3.0+ MB


In [39]:
df_clean.Sales = df_clean.Sales.astype('float')

In [40]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 64964 entries, 0 to 64963
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype   
---  ------     --------------  -----   
 0   Name       64964 non-null  object  
 1   Platform   64964 non-null  object  
 2   Year       64964 non-null  int64   
 3   Genre      64964 non-null  category
 4   Publisher  64964 non-null  object  
 5   Region     64964 non-null  object  
 6   Sales      64964 non-null  float64 
dtypes: category(1), float64(1), int64(1), object(4)
memory usage: 3.0+ MB


##To_CSV

In [41]:
df_clean.to_csv('vgames2_clean.csv',index=False)